src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -1281,27 +1267,25 @@ final int queueSize() {
          */
         final void push(ForkJoinTask<?> task, ForkJoinPool pool,
                         boolean internal) {
-            int s = top, b = base, m, cap, room; ForkJoinTask<?>[] a;
-            if ((a = array) != null && (cap = a.length) > 0) { // else disabled
-                if ((room = (m = cap - 1) - (s - b)) >= 0) {
-                    top = s + 1;
-                    long pos = slotOffset(m & s);
-                    if (!internal)
-                        U.putReference(a, pos, task);       // inside lock
-                    else
-                        U.getAndSetReference(a, pos, task); // fully fenced
-                    if (room == 0)                          // resize
-                        growArray(a, cap, s);
-                }
+            int s = top++, b = base, m, cap, room; ForkJoinTask<?>[] a;
+            if ((a = array) == null || (cap = a.length) <= 0 ||
+                (room = (m = cap - 1) - (s - b)) < 0) {
+                top = s;                          // back out
                 if (!internal)
                     unlockPhase();
-                if (room < 0)
-                    throw new RejectedExecutionException("Queue capacity exceeded");
-                else if ((room == 0 ||
-                          a[m & (s - 2)] == null) && // at most one existing task
-                         pool != null)
-                    pool.signalWork();
+                throw new RejectedExecutionException("Queue capacity exceeded");
             }
+            long k = slotOffset(m & s), pk = slotOffset(m & (s - 1));
+            if (!internal)
+                U.putReference(a, k, task);       // inside lock
+            else
+                U.getAndSetReference(a, k, task); // fully fenced
+            if (room == 0)                        // resize
+                growArray(a, cap, s);
+            if (!internal)
+                unlockPhase();
+            if ((room == 0 || U.getReference(a, pk) == null) && pool != null)
```

Comment:
We used to look at `- 2` but now we look at `- 1`, perhaps that could account for stalls?

Reply:
Well, the interplay of signal rules in push, runWorker, and deactivate lead to under/over signalling, scanning, contention. I'm about to commit a different tactic.

Reply:
Ok, staying tuned!